tests/unit/completion/test_models.py

```diff
@@ -498,3 +499,42 @@ def test_bind_completion(qtmodeltester, monkeypatch, stubs, config_stub,
             ('rock', "Alias for 'roll'", 'ro'),
         ]
     })
+
+
+def test_url_completion_benchmark(benchmark, config_stub,
+                                  quickmark_manager_stub,
+                                  bookmark_manager_stub,
+                                  web_history_stub):
+    """Benchmark url completion."""
+    config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
+                                      'web-history-max-items': 1000}
+
+    entries = [history.Entry(
+        atime=i,
+        url=QUrl('http://example.com/{}'.format(i)),
+        title='title{}'.format(i))
+        for i in range(100000)]
+
+    web_history_stub.history_dict = collections.OrderedDict(
+        ((e.url_str(), e) for e in entries))
+
+    quickmark_manager_stub.marks = collections.OrderedDict(
+        (e.title, e.url_str())
+        for e in entries[0:1000])
+
+    bookmark_manager_stub.marks = collections.OrderedDict(
+        (e.url_str(), e.title)
+        for e in entries[0:1000])
+
+    def bench():
+        model = urlmodel.UrlCompletionModel()
```

Comment:
In case @The-Compiler missed my comment on IRC: seems fine but i wonder if you could have the function even leaner. Do you need to make those model instances every time? it seems to me like the set_pattern is just side-effect free query method.